com.avaloq.tools.ddk.xtext.generator/src/com/avaloq/tools/ddk/xtext/generator/parser/antlr/AnnotationAwareXtextAntlrGeneratorFragment2.xtend

```diff
@@ -0,0 +1,218 @@
+/*******************************************************************************
+ * Copyright (c) 2016 Avaloq Evolution AG and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Avaloq Evolution AG - initial API and implementation
+ *******************************************************************************/
+
+package com.avaloq.tools.ddk.xtext.generator.parser.antlr
+
+import com.avaloq.tools.ddk.xtext.generator.parser.common.GrammarRuleAnnotations.SemanticPredicate
+import com.avaloq.tools.ddk.xtext.generator.parser.common.PredicatesNaming
+import com.avaloq.tools.ddk.xtext.parser.ISemanticPredicates
+import com.avaloq.tools.ddk.xtext.parser.antlr.AbstractContextualAntlrParser
+import com.avaloq.tools.ddk.xtext.parser.antlr.ParserContext
+import com.google.common.base.Joiner
+import com.google.inject.Inject
+import com.google.inject.Singleton
+import java.util.stream.Collectors
+import org.antlr.runtime.CharStream
+import org.antlr.runtime.Token
+import org.antlr.runtime.TokenSource
+import org.eclipse.xtext.parser.antlr.XtextTokenStream
+import org.eclipse.xtext.xtext.generator.grammarAccess.GrammarAccessExtensions
+import org.eclipse.xtext.xtext.generator.model.FileAccessFactory
+import org.eclipse.xtext.xtext.generator.model.JavaFileAccess
+import org.eclipse.xtext.xtext.generator.model.TypeReference
+import org.eclipse.xtext.xtext.generator.parser.antlr.GrammarNaming
+import org.eclipse.xtext.xtext.generator.parser.antlr.XtextAntlrGeneratorFragment2
+
+import static extension org.eclipse.xtext.GrammarUtil.*
+import static extension org.eclipse.xtext.xtext.generator.parser.antlr.AntlrGrammarGenUtil.*
+import com.avaloq.tools.ddk.xtext.generator.parser.common.GrammarRuleAnnotations
+
+class AnnotationAwareXtextAntlrGeneratorFragment2 extends XtextAntlrGeneratorFragment2 {
+
+  @Inject AnnotationAwareAntlrGrammarGenerator productionGenerator
+  @Inject GrammarNaming productionNaming
+  @Inject FileAccessFactory fileFactory
+
+  @Inject extension PredicatesNaming predicatesNaming
+  @Inject extension GrammarAccessExtensions grammarUtil
+  @Inject extension GrammarRuleAnnotations annotations
+
+  protected override void checkGrammar() {
+    super.checkGrammar();
+    getGrammar().annotateGrammar
+  }
+
+  protected override doGenerate() {
+      super.doGenerate()
+      generateAbstractSemanticPredicate().writeTo(projectConfig.runtime.srcGen)
+  }
+
+  protected override generateProductionGrammar() {

```

Comment:
Not sure I understand why this method has been overridden when the implementation is the same, and presumably all the correct classes will be injected into the super class.

Reply:
productionGenerator is the local field here and is injected again as the type is different. AnnotationAwareAntlrGrammarGenerator cannot be made a subtype of the type used in the parent class (AntlrGrammarGenerator) as we need to do super... calls to the super.super type.

Reply:
Ah OK. Makes sense.